1D Column Examples documentation for Heat and Advection Equations (#2297)
CliMA/ClimaCore.jl: docs/src/examples.md

# Examples

### Advection

The 1D Column advection example in [`examples/column/advect.jl`](https://github.com/CliMA/ClimaCore.jl/blob/main/examples/column/advect.jl).

#### Equations and Discretizations

> Follows the continuity equation

Reviewer: ```suggestion
Follows the advection equation
```

```math
\begin{equation}
  \frac{\partial \theta}{\partial t} = -\frac{\partial (v \theta)}{\partial z}
\label{eq:1d-column-advection-continuity}
\end{equation}
```
This is discretized using the following

```math
\begin{equation}

>   \frac{\partial \theta}{\partial t} \approx - D(V, \theta) 

Reviewer: Hi @onzerem , here on line 21, you used capital `V`, but below, on line 29 you used lower case `v`. Please be consistent.

\label{eq:1d-column-advection-discrete}
\end{equation}
```

#### Prognostic Variables

* ``\theta``: the scalar field
* ``v``: the velocity field

#### Tendencies 

The example code solves the equation for 4 different tendencies with the following discretizations:

> * Tendency 1: ``D = \partial(UB),`` where ``\partial`` is the [`face-to-center divergence`](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.DivergenceF2C) and $UB$ is the [`center-to-face upwind product operator`](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.UpwindBiasedProductC2F)

Reviewer: ```suggestion
* Tendency 1: ``D = \partial(UB),`` where ``\partial`` is the [`face-to-center divergence`](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.DivergenceF2C) and $UB$ is the [`center-to-face upwind biased product`](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.UpwindBiasedProductC2F) operator
```

> * Tendenct 2: Follows tendency 1 with the addition of flux correction `fcc`

Reviewer: Please describe in the formula and put a link to the `FluxCorrectionC2C` operator as you did for Tendency 1 here

> * Tendency 3: $D = A$, where $A$ is the [`discrete vertical advection`](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.AdvectionC2C)

Reviewer: ```suggestion
* Tendency 3: $D = A$, where $A$ is the [`center-to-center vertical advection`](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.AdvectionC2C) operator
```

> * Tendency 4: Follows tendency 3 with the addition of flux correction `fcc`

Reviewer: Same comment as for Tendency 2. Please spell out the formula and put a link for the operator used.

#### Set Up

> This test case is set up in a 1D column domain ``z \in [0, 4\pi]``. 

Reviewer: You should probably mention the boundary conditions here and the form of velocity field used, which is a sinusoidal wave. Thank you

### Heat

The 1D Column heat example in [`examples/column/heat.jl`](https://github.com/CliMA/ClimaCore.jl/blob/main/examples/column/heat.jl).

#### Equations and discretizations

> Follows the continuity equation

Reviewer: ```suggestion
Follows the heat equation
```

```math
\begin{equation}
  \frac{\partial T}{\partial t} = \alpha \cdot \nabla^2 T.
\label{eq:1d-column-heat-continuity}
\end{equation}
```

This is discretized using the following

```math
\begin{equation}
  \frac{\partial T}{\partial t} \approx \alpha \cdot D(G(T)).
\label{eq:1d-column-heat-discrete}
\end{equation}
```

#### Prognostic Variables

* ``\alpha``: thermal diffusivity measured in  $\frac{m^2}{s}$
* ``T``: temperature

#### Differentiation Operators

>  * ``D`` is the [face-to-center divergence](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.DivergenceF2C), called `divf2c` in the example code

Reviewer: ```suggestion
 * ``D`` is the [face-to-center divergence](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.DivergenceF2C) operator, called `divf2c` in the example code
```

>  * ``G`` is the [center-to-face gradient](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.GradientC2F), called `gradc2f` in the example code

Reviewer: ```suggestion
 * ``G`` is the [center-to-face gradient](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.GradientC2F) operator, called `gradc2f` in the example code
```

#### Set Up

> This test case is set up in a 1D column domain ``z \in [0, 1]``.

Reviewer: Please also mention the boundary conditions. 

From the code
```
bcs_bottom = Operators.SetValue(FT(0.0))
bcs_top = Operators.SetGradient(Geometry.WVector(FT(1.0)))
```

we can see that it has a Dirichelet (homogeneous) BC at the bottom and a Neumann boundary condition at the top. 
I believe. Is this correct, @dennisYatunin ?

# Examples

> ## 1D Column examples

Reviewer: Please put this header back in. 

```
## 1D Column examples
```

> ### Advection

Reviewer: ```suggestion
### Advection equation
```

The 1D Column advection example in [`examples/column/advect.jl`](https://github.com/CliMA/ClimaCore.jl/blob/main/examples/column/advect.jl).

#### Equations and Discretizations

Follows the advection equation

```math
\begin{equation}

>   \frac{\partial \theta}{\partial t} = -\frac{\partial (v \theta)}{\partial z}

Reviewer: Please use proper punctuation to connect or end sentences, even after math formulas/equations.

```suggestion
  \frac{\partial \theta}{\partial t} = -\frac{\partial (v \theta)}{\partial z} .
```

\label{eq:1d-column-advection-continuity}
\end{equation}
```
This is discretized using the following

```math
\begin{equation}
  \frac{\partial \theta}{\partial t} \approx - D(v, \theta) 
\label{eq:1d-column-advection-discrete}
\end{equation}
```

#### Prognostic Variables

* ``\theta``: the scalar field
* ``v``: the velocity field

#### Tendencies 

The example code solves the equation for 4 different tendencies with the following discretizations:

> * Tendency 1: ``D = \partial(UB)``, where ``\partial`` is the [`face-to-center divergence`](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.DivergenceF2C) and $UB$ is the [`center-to-face upwind biased product`](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.UpwindBiasedProductC2F) operator.
> * Tendency 2: Follows tendency 1, where ``D = \partial(UB)``. Similarly, ``\partial`` is the [`face-to-center divergence`](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.DivergenceF2C) and $UB$ is the [`center-to-face upwind biased product`](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.UpwindBiasedProductC2F) operator. Tendency 2 also includes the [`center-to-center flux correction`](https://github.com/CliMA/ClimaCore.jl/blob/main/src/Operators/finitedifference.jl#L2617), `fcc`.
> * Tendency 3: $D = A$, where $A$ is the [`center-to-center vertical advection`](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.AdvectionC2C) operator.
> * Tendency 4: Follows tendency 3, where $D = A$, the [`center-to-center vertical advection`](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.AdvectionC2C). Tendency 4 also includes the [`center-to-center flux correction`](https://github.com/CliMA/ClimaCore.jl/blob/main/src/Operators/finitedifference.jl#L2617), `fcc`.

Reviewer: I would reorganize this a bit better.

```suggestion
- Tendency 1:

  $$D = \partial(UB),$$  

  where ``\partial`` is the [`face-to-center divergence`](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.DivergenceF2C) and $UB$ is the [`center-to-face upwind biased product`](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.UpwindBiasedProductC2F) operator.
- Tendency 2:

  $$D = \partial(UB) + \textrm{fcc}(v, \theta),$$  
  
  where $\textrm{fcc}(v, \theta)$ is the [`center-to-center flux correction`](https://github.com/CliMA/ClimaCore.jl/blob/main/src/Operators/finitedifference.jl#L2617) operator.

- Tendency 3:

  $$D = A,$$  
   
  where $A$ is the [`center-to-center vertical advection`](https://clima.github.io/ClimaCore.jl/dev/operators/#ClimaCore.Operators.AdvectionC2C) operator.

- Tendency 4:

  $$D = A + \textrm{fcc}(v, \theta),$$  
  
  where $\textrm{fcc}(v, \theta)$ is the [`center-to-center flux correction`](https://github.com/CliMA/ClimaCore.jl/blob/main/src/Operators/finitedifference.jl#L2617) operator.
```

#### Set Up

This test case is set up in a 1D column domain ``z \in [0, 4\pi]``, discretized into a mesh of 128 elements. The velocity field is defined as a sinusoidal wave. The boundary conditions are operator dependent, so they depend on the tendency. 
* For tendencies 1 and 2 where the upwind biased operator ``UB`` is used, the left boundary is defined as ``sin(a - t)``. The right boundary is ``sin(b - t)``. Here ``a`` and ``b`` are the left and right bounds of the domain. 
* For tendencies 3 and 4, where the advection operator ``A`` is used, the left boundary is defined as ``sin(-t)``.  The right boundary is extrapolated, meaning its value is set to the closest interior point.

> ### Heat

Reviewer: ```suggestion
### Heat equation
```
